Refactoring of the parsing of the `--object-bits` command line argument

thomasspriggs opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

As I was working on #5440, I found the entry point code less than straight forward to follow. I carried out this refactorings locally in order to make it easier for me to understand. I am raising this separate PR containing the refactoring work only as it was not required in order to implement the feature in the original PR, but should still improve maintainability.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] N/A - No user visible changes ~The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/~
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [x] N/A - non claimed - ~My commit message includes data points confirming performance improvements (if claimed).~
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

thomasspriggs committed:
Add tests of the parsing of the `--object-bits` argument

In order to ensure that this functionality is not broken by subsequent
refactorings.

thomasspriggs committed:
Put default initialisation of `bv_encodingt` in its definition

So that `bv_encodingt` can be more used without the need to explicitly
initialise its members.

thomasspriggs committed:
Separate the parsing of `object-bits`

So the length of `configt::set` can be marginally reduced. So that this
functionality can be documented. This would also support unit testing
of this parsing if unit tests are added later.

thomasspriggs committed:
Add default base to `string2optional`

So that direct usage of this function can be more succinct.

thomasspriggs committed:
Use `parse_object_bits_encoding` in `parse_object_bits_encoding`

The previous use of `unsafe_string2unsigned` handled non numeric input
based on the behaviour of 0 being returned by `unsafe_string2unsigned`,
which is then caught by the subsequent range check. It was well defined
behaviour of `std::strtoul`. However this was only discoverable, based
on looking at the implementation of `unsafe_string2unsigned` and reading
the documentation of `std::strtoul` or by using a debugger. The explict
check in the refactored version makes the handling of this case clear
from reading `parse_object_bits_encoding` only, without reference to
other code or documentation.

hannes-steffenhagen-diffblue commented on src/util/config.cpp:
```diff
@@ -776,8 +776,8 @@ configt::bv_encodingt parse_object_bits_encoding(
   const std::size_t pointer_width)
 {
-  const unsigned int object_bits = unsafe_string2unsigned(argument);
-  if(object_bits <= 0 || object_bits >= pointer_width)
```
:pick: Would be better if the case where parsing fails and the `object_bits` are out of range were separated, also it would be helpful for debugging to put `argument` in the error message (this makes it easier to find if for example you've messed up a command line when copy&pasting, e.g. `--object-bits --function hello world 8`)

thomasspriggs replied:
Done.

hannes-steffenhagen-diffblue reviewed (approved):
Looks OK to me

hannes-steffenhagen-diffblue reviewed (approved):
Looks OK to me

smowton commented on regression/cbmc/object-bits-parsing/non_numeric.desc:
```diff
@@ -0,0 +1,8 @@
+^EXIT=1$
+^SIGNAL=0$
+--
+Test parsing of an invalid non numeric object-bits setting.
```
This is a must-not-match, not a comment (similarly others)

thomasspriggs replied:
Fixed.

smowton commented on src/util/config.cpp:
```diff
@@ -765,16 +765,38 @@ void configt::set_arch(const irep_idt &arch)
+  const std::size_t pointer_width)
+{
+  const auto object_bits = string2optional<unsigned int>(argument);
+  if(!object_bits || *object_bits <= 0 || *object_bits >= pointer_width)
```
Note `<= 0` on an unsigned type

thomasspriggs replied:
I have swapped this to use `== 0`, as this should be less misleading.

smowton reviewed (approved):
Couple of simple fixes otherwise lgtm

thomasspriggs committed:
Improve error messaging when invalid `--object-bits` is parsed

It is easier to debug your invocation of a command line tool if the
specific invalid argument and the reason why it is invalid is given.

thomasspriggs committed:
Avoid checking if an unsigned number is below 0

`object_bits <= 0` suggests that `object_bits` could be below zero. This
is not possible because `object_bits` is unsigned. Checking for equal to
zero makes it explicit that we are checking for one specific value.
